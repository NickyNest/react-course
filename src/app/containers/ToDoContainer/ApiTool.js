import fetch from 'isomorphic-fetch';

const endpoint = 'http://localhost:60253/api/tasks';

const fetchRequest = options => {
    const {method} = options;
    const {id} = options;
    const {body} = options;

    const url = id ? `${endpoint}/${id}` : endpoint;
    const init = {
        method,
        headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json'
        }
    };

    if (body) {
        init.body = JSON.stringify(body);
    }

    return fetch(url, init)
        .then(response => response.json())
        .then(responseData => responseData)
        .catch(error => { throw new Error(error); });
};

const get = () => fetchRequest({method: 'GET'});
const create = body => fetchRequest({method: 'POST', body});
const update = (id, body) => fetchRequest({method: 'PATCH', id, body});
const remove = id => fetchRequest({method: 'DELETE', id});

export {get, create, update, remove};