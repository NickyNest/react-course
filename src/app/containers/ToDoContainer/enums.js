const showModes = {
    showAll: 0,
    showCompleted: 1,
    showPending: 2
};

const sortModes = {
    None: 0,
    Up: 1,
    Down: 2
};

export {showModes, sortModes};