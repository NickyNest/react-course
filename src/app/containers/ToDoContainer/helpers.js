import faker from 'faker';

const toDate = weirdDate => {
    const parts = weirdDate.split('/');
    return new Date(parts[2], parts[0], parts[1]).toLocaleDateString();
};

const toWeirdDate = () => {
    const now = new Date();
    return `${now.getMonth()}/${now.getDate()}/${now.getFullYear()}`;
};

const buildTask = title => ({
    id: faker.random.uuid(),
    title,
    completed: false,
    createdDate: toWeirdDate()
});

export {toDate, buildTask};