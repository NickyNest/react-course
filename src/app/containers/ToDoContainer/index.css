.Todo-title {
    background-color: rgb(24, 141, 219);
    height: 30px;
    padding: 10px;
    color: white;
  }