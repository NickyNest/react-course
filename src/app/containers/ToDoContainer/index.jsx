import React, {Component} from 'react';
import TaskInput from 'components/TaskInput';
import Task from 'components/Task';
import FilterBar from 'components/FilterBar';
import {Grid} from 'semantic-ui-react';
import {SHOW_ALL, SORT_CREATED_DATE_NONE} from 'utils/actions';
import * as api from 'utils/taskStore';
import {toDate, buildTask, handleShowMode, handleSortMode} from 'utils/helpers';
import './index.css';

class ToDoContainer extends Component {
    constructor(props) {
        super(props);
        this.state = {
            tasks: [],
            showMode: SHOW_ALL,
            sortMode: SORT_CREATED_DATE_NONE
        };
    }

    componentWillMount() {
        this.getAllTasks();
    }

    getAllTasks = () => {
        api.get()
            .then(response => response.json())
            .then(response => {
                const tasks = response.map(task => ({...task, createdDate: toDate(task.createdDate)}));
                this.setState({tasks});
            });
    };

    addTask = title => {
        api.create(buildTask(title, new Date()))
            .then(response => response.json())
            .then(newTask =>
                this.setState({tasks: [...this.state.tasks, {...newTask, createdDate: toDate(newTask.createdDate)}] }));
    };

    updateTak = (id, isCompleted) => {
        api.update(id, {completed: isCompleted})
            .then(response => {
                if (response.ok) {
                    const {tasks} = this.state;
                    const taskIndex = tasks.findIndex(task => task.id === id);
                    const newTasks = [
                        ...tasks.slice(0, taskIndex),
                        {...tasks[taskIndex], completed: isCompleted},
                        ...tasks.slice(taskIndex + 1, taskIndex.length)
                    ];
                    this.setState({tasks: newTasks});
                } else {
                    throw new Error('Network response was not ok.');
                }
            });
    };

    removeTask = id => {
        api.remove(id)
            .then(response => {
                if (response.ok) {
                    this.setState({tasks: this.state.tasks.filter(task => task.id !== id)});
                } else {
                    throw new Error('Network response was not ok.');
                }
            });
    };

    mapToDoTaskList = tasks => (
        <div>
            {
                tasks.map(task =>
                    <Task key={task.id} task={task} onCheckChange={this.updateTak} onRemoveTask={this.removeTask} />)
            }
        </div>
    );

    changeShowMode = showMode => {
        this.setState({showMode});
    };

    removeCompleted = () => {
        this.state.tasks
            .filter(task => task.completed === true)
            .forEach(task => api.remove(task.id));

        this.getAllTasks();
    };

    changeSortMode = sortMode => {
        this.setState({sortMode});
    };

    render() {
        const {showMode} = this.state;

        const filteredTasks = handleShowMode(this.state.tasks, showMode);
        const filteredAndSortedTasks = handleSortMode(filteredTasks, this.state.sortMode);

        const title = 'Simple To-Do application';
        return (
            <div>
                <h3 className='Todo-title'>{title}</h3>
                <FilterBar
                    changeShowMode={this.changeShowMode}
                    removeCompleted={this.removeCompleted}
                    changeSortMode={this.changeSortMode}
                    currentMode={showMode} />
                <span>total: {filteredTasks.length}</span>
                <TaskInput onAddTask={this.addTask} />
                <Grid>
                    <Grid.Column width={5}>
                        {filteredAndSortedTasks.length === 0 ? 'huhÐª' : this.mapToDoTaskList(filteredAndSortedTasks)}
                    </Grid.Column>
                </Grid>
            </div>
        );
    }
}

export default ToDoContainer;