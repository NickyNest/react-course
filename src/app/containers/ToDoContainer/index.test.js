/* eslint-disable react/jsx-filename-extension */

import React from 'react';
import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import ToDoContainer from './index';

configure({adapter: new Adapter()});

describe('ToDoContainer', () => {
    it('ToDoContainer match expected snapshot', () => {
        expect(shallow(<ToDoContainer />)).toMatchSnapshot();
    });
});