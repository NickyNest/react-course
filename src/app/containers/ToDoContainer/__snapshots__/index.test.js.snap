// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoContainer ToDoContainer match expected snapshot 1`] = `
<div>
  <h3
    className="Todo-title"
  >
    Simple To-Do application
  </h3>
  <FilterBar
    changeShowMode={[Function]}
    changeSortMode={[Function]}
    currentMode="SHOW_ALL"
    removeCompleted={[Function]}
  />
  <span>
    total: 
    0
  </span>
  <TaskInput
    onAddTask={[Function]}
  />
  <Grid>
    <GridColumn
      width={5}
    >
      huhÐª
    </GridColumn>
  </Grid>
</div>
`;
